### PR TITLE
Edmund/rocksdb delete range

### PR DESCRIPTION
**Pull request by edmundnoble**
This was more complicated than I had hoped because I had to get a handle on the "default" column family to even do a DeleteRange, because it requires a column family as an argument.

### CODE REVIEW

**Comment by edmundnoble** [comment]
Turns out, by writing my own C++ shim, I could avoid a lot of complexity that was in this PR before.

File: chainweb-storage.cabal [hunk 1]
```diff
@@ -27,6 +27,11 @@ source-repository head
 
 library
     hs-source-dirs: src
+    cxx-sources: cpp/chainweb-rocksdb.cpp
+    includes: cpp/chainweb-rocksdb.h
+    install-includes: cpp/chainweb-rocksdb.h
+    extra-libraries: rocksdb stdc++
+    cxx-options: -std=c++11
```
Comment (larskuhtz) [inline chainweb-storage.cabal:34]: Is this a new build requirement? Or is c++11 included with GHC?

File: src/Data/CAS/RocksDB.hs [hunk 2]
```diff
@@ -785,61 +794,100 @@ foreign import ccall unsafe "rocksdb\\c.h rocksdb_checkpoint_create"
...
+  where
+    !range' = validateRangeOrdered table range
+    R.DB dbPtr _ = _rocksDbTableDb table
+
+-- | Read a value by key.
+-- One less copy than the version in rocksdb-haskell by using unsafePackCStringFinalizer.
+get :: MonadIO m => R.DB -> R.ReadOptions -> ByteString -> m (Maybe ByteString)
+get (R.DB db_ptr _) opts key = liftIO $ R.withCReadOpts opts $ \opts_ptr ->
+    BU.unsafeUseAsCStringLen key $ \(key_ptr, klen) ->
+    alloca                       $ \vlen_ptr -> do
```
Comment (larskuhtz) [inline src/Data/CAS/RocksDB.hs]:
```suggestion
    alloca $ \vlen_ptr -> do
```

File: src/Data/CAS/RocksDB.hs [hunk 3]
```diff
@@ -785,61 +794,100 @@ foreign import ccall unsafe "rocksdb\\c.h rocksdb_checkpoint_create"
...
+            maxKeyPtr (fromIntegral maxKeyLen :: CSize)
+  where
+    !range' = validateRangeOrdered table range
+    R.DB dbPtr _ = _rocksDbTableDb table
+
+-- | Read a value by key.
+-- One less copy than the version in rocksdb-haskell by using unsafePackCStringFinalizer.
+get :: MonadIO m => R.DB -> R.ReadOptions -> ByteString -> m (Maybe ByteString)
+get (R.DB db_ptr _) opts key = liftIO $ R.withCReadOpts opts $ \opts_ptr ->
+    BU.unsafeUseAsCStringLen key $ \(key_ptr, klen) ->
```
Comment (larskuhtz) [inline src/Data/CAS/RocksDB.hs:882]: Why is it possible to use `unsafeUseAsCStringLen` here and not for the `range` values above?

File: src/Data/CAS/RocksDB.hs [hunk 4]
```diff
@@ -785,61 +794,100 @@ foreign import ccall unsafe "rocksdb\\c.h rocksdb_checkpoint_create"
...
+    )
+
+-- | Batch delete a range of keys in a table. 
+-- Throws if the range of the *encoded keys* is not ordered (lower, upper).
+deleteRangeRocksDb :: HasCallStack => RocksDbTable k v -> (Maybe k, Maybe k) -> IO ()
+deleteRangeRocksDb table range = do
+    let !range' = validateRangeOrdered table range
+    let R.DB dbPtr _ = _rocksDbTableDb table
+    R.withCWriteOpts R.defaultWriteOptions $ \optsPtr ->
+        B.useAsCStringLen (fst range') $ \(minKeyPtr, minKeyLen) ->
```
Comment (larskuhtz) [inline src/Data/CAS/RocksDB.hs]: Would it be safe to use `unsafeUseAsCStringLen` here, too?

**Review by larskuhtz — APPROVED** [review]
I have one suggestion and a question. Looks good otherwise.

File: chainweb-storage.cabal (hunk 1 above)
Comment (edmundnoble) [reply chainweb-storage.cabal:34]: I believe this is a new build requirement. On my system, ghcup reports that the following are all of the requirements for GHC:
`build-essential curl libffi-dev libffi7 libgmp-dev libgmp10 libncurses-dev libncurses5 libtinfo5`
And I don't think GHC comes with GCC & libstdc++. 
Added something to the chainweb-storage README about this.

File: src/Data/CAS/RocksDB.hs (hunk 3 above)
Comment (edmundnoble) [reply src/Data/CAS/RocksDB.hs:882]: I see no reason we couldn't, either way rocksdb will not alter the contents of either pointer. Fixed.

File: src/Data/CAS/RocksDB.hs [hunk 5]
```diff
@@ -785,61 +794,100 @@ foreign import ccall unsafe "rocksdb\\c.h rocksdb_checkpoint_create"
...
+        B.useAsCStringLen (snd range') $ \(maxKeyPtr, maxKeyLen) ->
```
Comment (larskuhtz) [inline src/Data/CAS/RocksDB.hs]:
```suggestion
        B.unsafeUseAsCStringLen (fst range') $ \(minKeyPtr, minKeyLen) ->
        B.unsafeUseAsCStringLen (snd range') $ \(maxKeyPtr, maxKeyLen) ->
```

Comment (larskuhtz) [reply src/Data/CAS/RocksDB.hs]: ... assuming that `range'` isn't changed by rocksDb

**Review by larskuhtz — COMMENTED** [review]
One more suggestion to use `unsafeUseAsCString`

**Pull request merged**